storage_controller/src/service.rs

```diff
@@ -4442,6 +4457,306 @@ impl Service {
+                    self.config
+                        .safekeeper_jwt_token
+                        .clone()
+                        .map(SecretString::from),
+                );
+
+                let _res = client
+                    .bump_timeline_term(
+                        tenant_id,
+                        timeline_id,
+                        &TimelineTermBumpRequest { term: Some(0) }, // TODO
+                    )
+                    .await
+                    .map_err(|e| ApiError::BadRequest(anyhow::anyhow!(e)))?;
+            }
+        }
+
+        // 7. Repeatedly call PUT configuration on safekeepers from the new set, delivering them joint_conf and collecting their positions.
+
+        for sk in safekeepers.values() {
+            if joint_config
+                .new_members
+                .as_ref()
+                .unwrap()
+                .contains(sk.get_id())
+            {
+                let client = SafekeeperClient::new(
+                    sk.get_id(),
+                    self.http_client.clone(),
+                    sk.base_url(),
+                    self.config
+                        .safekeeper_jwt_token
+                        .clone()
+                        .map(SecretString::from),
+                );
+
+                let _res = client
+                    .switch_timeline_membership(
+                        tenant_id,
+                        timeline_id,
+                        &TimelineMembershipSwitchRequest {
+                            mconf: joint_config.clone(),
+                        },
+                    )
+                    .await
+                    .map_err(|e| ApiError::BadRequest(anyhow::anyhow!(e)))?;
+            }
+        }
+
+        // 8. Create new_conf: Configuration incrementing join_conf generation and having new safekeeper set as sk_set and None new_sk_set.
+
+        let new_conf = membership::Configuration {
+            generation: generation.next(),
+            members: desired_member_set,
+            new_members: None,
+        };
+
+        self.persistence
+            .update_timeline_membership(
+                tenant_id,
+                timeline_id,
+                new_conf.generation,
+                &desired_sk_ids,
+                None,
+            )
+            .await?;
+
+        // 9. Call PUT configuration on safekeepers from the new set, delivering them new_conf.
+
+        for sk in safekeepers.values() {
+            if new_conf.members.contains(sk.get_id()) {
+                let client = SafekeeperClient::new(
+                    sk.get_id(),
+                    self.http_client.clone(),
+                    sk.base_url(),
+                    self.config
+                        .safekeeper_jwt_token
+                        .clone()
+                        .map(SecretString::from),
+                );
+
+                let _res = client
+                    .switch_timeline_membership(
```

Comment:
joinset again, plus putting the unsuccessful attempts into the reconciler. if we didn't reconcile some op to some safekeeper it's fine to overwrite it.